server/src/main/java/org/apache/iotdb/db/utils/QueryDataSetUtils.java

```diff
@@ -174,6 +180,126 @@ public static TSQueryDataSet convertQueryDataSetByFetchSize(QueryDataSet queryDa
+                valueOccupation[k] += 8;
+                break;
+              case FLOAT:
+                dataOutputStream.writeFloat(field.getFloatV());
+                valueOccupation[k] += 4;
+                break;
+              case DOUBLE:
+                dataOutputStream.writeDouble(field.getDoubleV());
+                valueOccupation[k] += 8;
+                break;
+              case BOOLEAN:
+                dataOutputStream.writeBoolean(field.getBoolV());
+                valueOccupation[k] += 1;
+                break;
+              case TEXT:
+                dataOutputStream.writeInt(field.getBinaryV().getLength());
+                dataOutputStream.write(field.getBinaryV().getValues());
+                valueOccupation[k] = valueOccupation[k] + 4 + field.getBinaryV().getLength();
+                break;
+              default:
+                throw new UnSupportedDataTypeException(
+                    String.format("Data type %s is not supported.", type));
+            }
+          }
+        }
+        rowCount++;
+        if (rowCount % 8 == 0) {
+          for (int j = 0; j < bitmap.length; j++) {
+            DataOutputStream dataBitmapOutputStream = dataOutputStreams[3*j + 2];
+            dataBitmapOutputStream.writeByte(bitmap[j]);
+            // we should clear the bitmap every 8 row record
+            bitmap[j] = 0;
+          }
+        }
+      } else {
+        break;
+      }
+    }
+
+    // feed the remaining bitmap
+    int remaining = rowCount % 8;
+    if (remaining != 0) {
+      for (int j = 0; j < bitmap.length; j++) {
+        DataOutputStream dataBitmapOutputStream = dataOutputStreams[3*j + 2];
+        dataBitmapOutputStream.writeByte(bitmap[j] << (8-remaining));
+      }
+    }
+
+    // calculate the time buffer size
+    int timeOccupation = rowCount * 8;
+
+    // calculate the bitmap buffer size
+    int bitmapOccupation = rowCount / 8 + (rowCount % 8 == 0 ? 0 : 1);
+    List<ByteBuffer> timeList = new LinkedList<>();
+    List<ByteBuffer> bitmapList = new LinkedList<>();
+    List<ByteBuffer> valueList = new LinkedList<>();
+    for (int i = 0; i < byteArrayOutputStreams.length; i += 3) {
+      ByteBuffer timeBuffer = ByteBuffer.allocate(timeOccupation);
+      timeBuffer.put(byteArrayOutputStreams[i].toByteArray());
+      timeBuffer.flip();
+      timeList.add(timeBuffer);
+      
+      ByteBuffer valueBuffer = ByteBuffer.allocate(valueOccupation[i/3]);
+      valueBuffer.put(byteArrayOutputStreams[i+1].toByteArray());
+      valueBuffer.flip();
+      valueList.add(valueBuffer);
+
+      ByteBuffer bitmapBuffer = ByteBuffer.allocate(bitmapOccupation);
+      bitmapBuffer.put(byteArrayOutputStreams[i+2].toByteArray());
+      bitmapBuffer.flip();
+      bitmapList.add(bitmapBuffer);
+    }
+    tsQueryNonAlignDataSet.setBitmapList(bitmapList);
+    tsQueryNonAlignDataSet.setValueList(valueList);
+    return tsQueryNonAlignDataSet;
```

Comment:
[Beyond this PR]
There will be 3 ByteBuffers for each timeseries (or column). When the number of columns is large while the fetch size is relatively small, this can introduce structural overhead that may not be negligible.
One way to reduce the overhead may be putting them all together in a ByteBuffer and record the offsets and lengths of each sub-buffer.
This is not a job for now but it seems worth trying.

Reply:
If you put them into one ByteBuffer, the multithread will lose its advantages.

Reply:
I mean to do it in the return phase (just before sending the result to the client), where I think there is no multithread.

Reply:
I find out this part of code has lots of bugs and it is useless.  Removed.